File: src/api/ws/router.py
import asyncio
import time

import orjson
import structlog
from fastapi import APIRouter
from starlette.websockets import WebSocket, WebSocketDisconnect

from config.provider import GlobalConfigProvider
from server.middleware.auth import (_get_static_key_context,
                                    _parse_bearer_token, feature_in_scope)

from .connection_manager import conn_mgr

logger = structlog.get_logger("ws.router")
router = APIRouter(tags=["WebSocket"])


async def _authenticate(websocket: WebSocket) -> dict | None:
    """
    Authenticates the WebSocket connection via the first JSON message.
    Expected: { "type": "auth", "token": "<base64(key_name:secret)>" }
    Closes with code 4001/4003 on failure.
    """
    config = GlobalConfigProvider().get_config()
    try:
        raw = await asyncio.wait_for(
            websocket.receive_text(), timeout=config.websocket.auth_timeout
        )
        msg = orjson.loads(raw)
    except asyncio.TimeoutError:
        await websocket.close(code=4001, reason="Auth timeout")
        return None
    except Exception:
        await websocket.close(code=4001, reason="Invalid JSON in auth message")
        return None

    if msg.get("type") != "auth" or not msg.get("token"):
        await websocket.close(code=4001, reason="First message must be auth")
        return None

    # Reuse the existing bearer token parser — identical to REST auth
    from fastapi.security import HTTPAuthorizationCredentials

    try:
        creds = HTTPAuthorizationCredentials(scheme="Bearer", credentials=msg["token"])
        key_name, secret = _parse_bearer_token(creds)
        auth_ctx = _get_static_key_context(key_name, secret, config)
    except Exception as exc:
        await websocket.close(code=4003, reason=str(exc))
        return None

    if not feature_in_scope("ws", auth_ctx):
        await websocket.close(
            code=4003,
            reason="API key does not have permission to use the WebSocket subsystem.",
        )
        return None

    return {
        "client_id": f"{auth_ctx.api_key_name}_{id(websocket)}",
        "api_key_name": auth_ctx.api_key_name,
        "db_scope": auth_ctx.db_scope,
        "fs_scope": auth_ctx.fs_scope,
    }


@router.websocket("")
async def ws_endpoint(websocket: WebSocket) -> None:
    """
    Primary WebSocket endpoint. Lifecycle:
      1. Accept raw socket upgrade.
      2. Wait for auth message (5s timeout).
      3. Register in ConnectionManager with API key scopes.
      4. Dispatch heartbeat task + message loop concurrently.
      5. Disconnect and clean up on close or error.
    """
    config = GlobalConfigProvider().get_config()
    if conn_mgr.active_count >= config.websocket.max_connections:
        await websocket.close(code=1013, reason="Server at maximum capacity")
        return

    await websocket.accept()

    auth = await _authenticate(websocket)
    if not auth:
        return

    client_id = auth["client_id"]
    scopes = {"db_scope": auth["db_scope"], "fs_scope": auth["fs_scope"]}

    # Register connection (do NOT accept again — already accepted above)
    await conn_mgr.register_pre_accepted(client_id, websocket, scopes)
    logger.info("WebSocket authenticated", client_id=client_id)

    # Send confirmation
    await websocket.send_text(
        orjson.dumps(
            {
                "type": "connected",
                "client_id": client_id,
            }
        ).decode("utf-8")
    )

    async def _heartbeat() -> None:
        heartbeat_interval = (
            GlobalConfigProvider().get_config().websocket.heartbeat_interval
        )
        while True:
            await asyncio.sleep(heartbeat_interval)
            try:
                await websocket.send_text(
                    orjson.dumps(
                        {
                            "type": "heartbeat",
                            "server_time": time.strftime(
                                "%Y-%m-%dT%H:%M:%SZ", time.gmtime()
                            ),
                        }
                    ).decode("utf-8")
                )
            except Exception:
                break

    heartbeat_task = asyncio.create_task(_heartbeat())

    try:
        while True:
            raw = await websocket.receive_text()
            try:
                msg = orjson.loads(raw)
            except Exception:
                continue

            msg_type = msg.get("type")

            if msg_type == "subscribe":
                topic = msg.get("topic", "")
                ok = conn_mgr.subscribe(client_id, topic)
                await websocket.send_text(
                    orjson.dumps(
                        {
                            "type": "ack",
                            "request_id": msg.get("request_id", ""),
                            "status": "ok" if ok else "denied",
                            "topic": topic,
                        }
                    ).decode("utf-8")
                )

            elif msg_type == "unsubscribe":
                topic = msg.get("topic", "")
                conn_mgr.unsubscribe(client_id, topic)

            elif msg_type == "pong":
                pass  # Client acknowledging our heartbeat

    except WebSocketDisconnect:
        logger.info("WebSocket disconnected", client_id=client_id)
    except Exception as exc:
        logger.warning("WebSocket error", client_id=client_id, error=str(exc))
    finally:
        heartbeat_task.cancel()
        conn_mgr.disconnect(client_id)

File: src/api/ws/router.rs
use axum::{
    extract::ws::{Message, WebSocket, WebSocketUpgrade},
    response::Response,
    routing::get,
    Router,
};
use futures::{sink::SinkExt, stream::StreamExt};
use serde_json::json;
use std::time::Duration;
use tokio::time::sleep;

use crate::api::ws::connection_manager::{ClientScopes, CONN_MGR};
use crate::config::loader::ConfigManager;

use crate::utils::types::AuthContext;

pub fn get_router() -> Router {
    Router::new().route("/", get(ws_handler))
}

pub async fn ws_handler(
    ws: WebSocketUpgrade,
    auth: Option<axum::extract::Extension<AuthContext>>,
) -> Response {
    let auth = auth.map(|axum::extract::Extension(a)| a);
    ws.on_upgrade(move |socket| handle_socket(socket, auth))
}

async fn handle_socket(socket: WebSocket, initial_auth: Option<AuthContext>) {
    let (mut sender, mut receiver) = socket.split();

    let config = ConfigManager::get();

    // 1. Authenticate if not already authenticated via headers
    let auth = if let Some(a) = initial_auth {
        a
    } else {
        let auth_timeout = config.websocket.auth_timeout;
        let auth_result = tokio::time::timeout(Duration::from_secs_f64(auth_timeout), async {
            while let Some(msg) = receiver.next().await {
                match msg {
                    Ok(Message::Text(text)) => {
                        if let Ok(json) = serde_json::from_str::<serde_json::Value>(&text) {
                            if json["type"] == "auth" {
                                if let Some(token) = json["token"].as_str() {
                                    let auth_value = if token.starts_with("Bearer ") {
                                        token.to_string()
                                    } else {
                                        format!("Bearer {}", token)
                                    };

                                    if let Some(ctx) = crate::middleware::auth::validate_api_key(
                                        &auth_value,
                                        &config,
                                    ) {
                                        return Ok(ctx);
                                    }
                                }
                                return Err("INVALID_TOKEN");
                            }
                        }
                        return Err("EXPECTED_AUTH_MESSAGE");
                    }
                    _ => return Err("INVALID_MESSAGE_TYPE"),
                }
            }
            Err("CONNECTION_CLOSED")
        })
        .await;

        match auth_result {
            Ok(Ok(ctx)) => ctx,
            _ => {
                let code = match auth_result {
                    Ok(Err(e)) => e,
                    _ => "AUTH_TIMEOUT",
                };
                let error = json!({
                    "type": "error",
                    "code": code,
                    "message": "Authentication failed or timed out"
                })
                .to_string();
                let _ = sender.send(Message::Text(error)).await;
                return;
            }
        }
    };

    let client_id = format!("{}_{}", auth.api_key_name, uuid::Uuid::new_v4());
    let mut scopes = ClientScopes::default();
    scopes.db_scope = auth.db_scope.clone();
    scopes.fs_scope = auth.fs_scope.clone();

    // 2. Setup internal communication channel
    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<Message>();
    CONN_MGR.register(&client_id, tx, scopes).await;

    // Send connection success
    let ack = json!({
        "type": "connected",
        "client_id": client_id
    })
    .to_string();
    let _ = sender.send(Message::Text(ack)).await;

    // 3. Heartbeat task
    let heartbeat_interval = config.websocket.heartbeat_interval as u64;
    let hb_client_id = client_id.clone();
    let hb_task = tokio::spawn(async move {
        loop {
            sleep(Duration::from_secs(heartbeat_interval)).await;
            let hb = json!({
                "type": "heartbeat",
                "server_time": chrono::Utc::now().to_rfc3339()
            })
            .to_string();
            CONN_MGR.send(&hb_client_id, hb).await;
        }
    });

    // 4. Multiplex internal messages to the WebSocket
    let mut send_task = tokio::spawn(async move {
        while let Some(msg) = rx.recv().await {
            if sender.send(msg).await.is_err() {
                break;
            }
        }
    });

    // 5. Read messages from the WebSocket
    let rc_client_id = client_id.clone();
    let mut recv_task = tokio::spawn(async move {
        while let Some(Ok(Message::Text(text))) = receiver.next().await {
            if let Ok(json) = serde_json::from_str::<serde_json::Value>(&text) {
                if let Some(msg_type) = json["type"].as_str() {
                    match msg_type {
                        "subscribe" => {
                            if let Some(topic) = json["topic"].as_str() {
                                let ok = CONN_MGR.subscribe(&rc_client_id, topic).await;
                                let ack = json!({
                                    "type": "ack",
                                    "request_id": json["request_id"].as_str().unwrap_or(""),
                                    "status": if ok { "ok" } else { "denied" },
                                    "topic": topic
                                })
                                .to_string();
                                CONN_MGR.send(&rc_client_id, ack).await;
                            }
                        }
                        "unsubscribe" => {
                            if let Some(topic) = json["topic"].as_str() {
                                CONN_MGR.unsubscribe(&rc_client_id, topic).await;
                            }
                        }
                        "pong" => {} // Heartbeat ack
                        _ => {}
                    }
                }
            }
        }
    });

    // Wait for either the sender or receiver task to die
    tokio::select! {
        _ = (&mut send_task) => recv_task.abort(),
        _ = (&mut recv_task) => send_task.abort(),
    };

    // 6. Cleanup
    hb_task.abort();
    CONN_MGR.disconnect(&client_id).await;
}
